### ISSUE TITLE
NullPointerException when fetching unnamed nested records in plain SQL with PostgreSQL

### ISSUE THREAD

**Issue by Conorrr** [jOOQ/jOOQ#6404]
Hi,

I am getting a null pointer exception when I attempt to call a function that returns a single record.

Example postgres function:
```
create function npeTest()
    returns record
    as $$
    declare
    v_return record;
    begin
    select 1 into v_return;
              return v_return;
    end;
    $$ language plpgsql;
```

Set sql dialect to Postgres
running:
```
 jooq.fetchOne("select npeTest()");
```

causes 

```
java.lang.NullPointerException: null
	at org.jooq.impl.Fields.size(Fields.java:83) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.RowImpl.size(RowImpl.java:227) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.AbstractRecord.<init>(AbstractRecord.java:136) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.RecordImpl.<init>(RecordImpl.java:134) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.Tools$1.newInstance(Tools.java:584) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.Tools.newRecord(Tools.java:558) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.Tools.newRecord(Tools.java:550) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.Tools.newRecord(Tools.java:514) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.DefaultBinding.pgNewRecord(DefaultBinding.java:2365) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.DefaultBinding.get(DefaultBinding.java:1595) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1653) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1624) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1589) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:128) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchOne(CursorImpl.java:1553) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1520) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:342) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.CursorImpl.fetchOne(CursorImpl.java:300) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.Tools.fetchOne(Tools.java:1486) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.AbstractResultQuery.fetchOne(AbstractResultQuery.java:520) ~[jooq-3.9.3.jar:na]
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:860) ~[jooq-3.9.3.jar:na]
```

Thanks

Conor

**Comment by lukaseder** [comment]
Thank you very much for your report. Interesting, indeed jOOQ currently doesn't handle anonymous nested records correctly. It is also not possible right now to `SELECT ROW(1)`. There's a pending feature request for this: https://github.com/jOOQ/jOOQ/issues/2360

That issue is for adding support for nested records in the DSL. It certainly makes sense to support nested records already through the plain SQL API. The expected implied result type in your case would be a `Record1<Record1<Integer>>`. I'll see if this can be discovered from a JDBC `ResultSetMetaData`

As a workaround, you could use the hstore extension to manually unnest your record:

```sql
SELECT (avals(hstore(npeTest())))[1]
```

**Comment by Conorrr** [comment]
Thanks for the response.

To work around it I am manually unwrapping the record. 

```sql
SELECT * FROM npeTest() AS (col integer)
```

**Comment by lukaseder** [comment]
Ah, yes indeed. I keep forgetting about the possibility of providing types with derived column lists. That's probably the soundest solution, also from a SQL perspective, if you plan on using that function in more complex contexts.

**Comment by lukaseder** [comment]
This is fixed for 3.10 now, and will be merged to 3.9.4. However, I still think that your solution might be better because type information will be explicit and can be picked up by jOOQ correctly. When using the `SELECT npeTest()` syntax, jOOQ will deserialise everything as `String` as the PostgreSQL JDBC driver doesn't communicate any type information for nested rows...

**Comment by lukaseder** [comment]
Related issue: #7691
